classes/migration/upgrade/v3_4_0/I8073_RemoveNotesWithoutQueriesAndRelatedObjects.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+/**
+ * @file classes/migration/upgrade/v3_4_0/I8073_RemoveNotesWithoutQueriesAndRelatedObjects.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class I8073_RemoveNotesWithoutQueriesAndRelatedObjects
+ * @brief Removes Notes without Queries and related objects
+ */
+
+namespace PKP\migration\upgrade\v3_4_0;
+
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+use PKP\migration\Migration;
+use Illuminate\Support\Facades\DB;
+use PKP\config\Config;
+use League\Flysystem\Local\LocalFilesystemAdapter;
+use League\Flysystem\UnixVisibility\PortableVisibilityConverter;
+use League\Flysystem\Filesystem;
+use League\Flysystem\FilesystemException;
+use League\Flysystem\UnableToDeleteFile;
+
+class I8073_RemoveNotesWithoutQueriesAndRelatedObjects extends Migration
+{
+    private const ASSOC_TYPE_NOTE = 0x0000208; // PKPApplication::ASSOC_TYPE_NOTE
+    private const ASSOC_TYPE_QUERY = 0x010000a; // PKPApplication::ASSOC_TYPE_QUERY
+    private const SUBMISSION_FILE_QUERY = 18; // SubmissionFile::SUBMISSION_FILE_QUERY
+    private const FILE_MODE_MASK = 0666; // FileManager::FILE_MODE_MASK
+    private const DIRECTORY_MODE_MASK = 0777; // FileManager::DIRECTORY_MODE_MASK
+
+    public function up(): void
+    {
+        // Does not have the foreign key reference
+        Schema::table('notification_settings', function (Blueprint $table) {
+            $table->foreign('notification_id')->references('notification_id')->on('notifications')->onDelete('cascade');
```

Comment:
Wherever you're adding foreign key constraints, you'll also need to clear out any bad data beforehand. See `lib/pkp/classes/migration/upgrade/v3_4_0/PreflightCheckMigration.php` for lots of examples. You can add new cleanups there.

Reply:
Clean-up code added